drivers/stepper/adi/tmc5041_stepper_controller.c

```diff
@@ -0,0 +1,608 @@
+	}
+
+	*velocity_fclk =
+		((uint64_t)(velocity_hz) << TMC5041_CLOCK_FREQ_SHIFT) / config->clock_frequency;
+	LOG_DBG("Stepper motor controller %s velocity: %d Hz, velocity_fclk: %d", dev->name,
+		velocity_hz, *velocity_fclk);
+}
+
+static int set_async_signal(const struct device *dev, struct k_poll_signal *async)
+{
+	struct tmc5041_stepper_data *data = dev->data;
+
+	if (async) {
+#ifdef CONFIG_STEPPER_ADI_TMC5041_INT
+		if (data->async_signal) {
+			k_poll_signal_reset(data->async_signal);
+		}
+		data->async_signal = async;
+#else
+		ARG_UNUSED(data);
+		LOG_WRN("STEPPER_ADI_TMC5041_INT not enabled, async signal not supported");
+#endif
+	}
+	return 0;
+}
+
+#ifdef CONFIG_STEPPER_ADI_TMC_STALLGUARD
+static int32_t print_stall_status(const struct device *dev)
+{
+	const struct tmc5041_stepper_config *config = dev->config;
+
+	int32_t sg_result = 0;
+
+	(void)tmc5041_read(config->controller, TMC5041_DRVSTATUS(config->index), &sg_result);
+
+	LOG_DBG("Detecting Load for motor %s %ld %ld", dev->name,
+		sg_result & TMC5041_DRV_STATUS_SG_RESULT_MASK,
+		(sg_result & TMC5041_DRV_STATUS_SG_STATUS_MASK) >>
+			TMC5041_DRV_STATUS_SG_STATUS_SHIFT);
+
+	return 0;
+}
+
+static int stallguard_enable(const struct device *dev, const bool enable)
+{
+	const struct tmc5041_stepper_config *config = dev->config;
+	uint32_t reg_value = 0;
+	(void)tmc5041_read(config->controller, TMC5041_SWMODE(config->index), &reg_value);
+	if (enable) {
+		reg_value |= TMC5041_SW_MODE_SG_STOP_ENABLE;
+	} else {
+		reg_value &= ~TMC5041_SW_MODE_SG_STOP_ENABLE;
+	}
+	(void)tmc5041_write(config->controller, TMC5041_SWMODE(config->index), reg_value);
+	return 0;
+}
+
+static void stallguard_work_handler(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct tmc5041_stepper_data *data =
+		CONTAINER_OF(dwork, struct tmc5041_stepper_data, stallguard_dwork);
+	LOG_DBG("Stallguard work handler");
+	print_stall_status(data->stepper);
+	stallguard_enable(data->stepper, true);
+}
+#endif
+
+#ifdef CONFIG_STEPPER_ADI_TMC5041_INT
+static void int_pin_callback_handler(const struct device *port, struct gpio_callback *cb,
+				     gpio_port_pins_t pins)
+{
+	ARG_UNUSED(port);
+	ARG_UNUSED(pins);
+	struct tmc5041_data *data = CONTAINER_OF(cb, struct tmc5041_data, int_cb);
+	(void)k_work_submit(&(data->int_pin_callback_work));
+}
+
+static void callback_work_handler(struct k_work *work)
+{
+	struct tmc5041_data *data = CONTAINER_OF(work, struct tmc5041_data, int_pin_callback_work);
+
+	if (data->controller == NULL) {
+		LOG_ERR("Controller not initialized");
+		return;
+	}
+	const struct tmc5041_config *config = data->controller->config;
+
+	for (uint8_t idx = 0; idx < config->motor_count; idx++) {
+		struct tmc5041_stepper_data *stepper_data = config->motors[idx]->data;
+
+		uint32_t drv_status;
+
+		(void)tmc5041_read(data->controller, TMC5041_DRVSTATUS(idx), &drv_status);
+		if (1U == FIELD_GET(TMC5041_DRVSTATUS_SG_MASK, drv_status)) {
```

Comment:
eeh, I think I know why but that's a bit uncommon in the code base :-) how about a more traditional

```c
if (FIELD_GET(TMC5041_DRVSTATUS_SG_MASK, drv_status) == 1U) {
```